src/oic/utils/session_backend.py

```diff
@@ -0,0 +1,201 @@
+
+    def get_verified_logout(self, uid: str) -> Optional[str]:
+        """Return logout verification key for given uid."""
+        # Since all the sessions should be the same, we return the first one
+        sids = self.get_by_uid(uid)
+        if len(sids) == 0:
+            return None
+        _dict = self[sids[0]]
+        if "verified_logout" not in _dict:
+            return None
+        return _dict["verified_logout"]
+
+    def get_token_ids(self, uid: str) -> List[str]:
+        """Return id_tokens for the given uid."""
+        return [self[sid]["id_token"] for sid in self.get_by_uid(uid)]
+
+    def is_revoke_uid(self, uid: str) -> bool:
+        """Return if the session is revoked."""
+        # We do not care which session it is - once revoked, al are revoked
+        return any([self[sid]["revoked"] for sid in self.get_by_uid(uid)])
+
+    def update(self, key, attribute, value):
+        """
+        Updates information stored. If the key is not know a new entry will be
+        constructed.
+
+        :param key: Key to the database
+        :param attribute: Attribute name
+        :param value: Attribute value
+        """
+        if key not in self:
+            self[key] = {attribute: value}
+        else:
+            item = self[key]
+            item[attribute] = value
+            self[key] = item
+
+
+class DictSessionBackend(SessionBackend):
+    """
+    Simple implementation of `SessionBackend` based on dictionary.
+
+    This should really not be used in production.
+    """
+
+    def __init__(self):
+        """Create the storage."""
+        self.storage = {}  # type: Dict[str, Dict[str, str]]
+
+    def __setitem__(self, key: str, value: Dict[str, str]) -> None:
+        """Store the session info in the storage."""
+        self.storage[key] = value
+
+    def __getitem__(self, key: str) -> Dict[str, str]:
+        """Retrieve session information based on session id."""
+        return self.storage[key]
+
+    def __delitem__(self, key: str) -> None:
+        """Delete the session info."""
+        del self.storage[key]
+
+    def __contains__(self, key: str) -> bool:
+        return key in self.storage
+
+    def get_by_sub(self, sub: str) -> List[str]:
+        """Return session ids based on sub."""
+        return [
+            sid for sid, session in self.storage.items() if session.get("sub") == sub
+        ]
+
+    def get_by_uid(self, uid: str) -> List[str]:
+        """Return session ids based on uid."""
+        return [
+            sid
+            for sid, session in self.storage.items()
+            if AuthnEvent.from_json(session["authn_event"]).uid == uid
+        ]
+
+    def get(self, attr: str, val: str) -> List[str]:
+        """Return session ids based on attribute name and value."""
+        return [
+            sid for sid, session in self.storage.items() if session.get(attr) == val
+        ]
+
+    def get_uid_by_sub(self, sub):
```

Comment:
These methods need to be moved to the `SessionBackend`, otherwise the `SessionDB` wouldn't work with other implementations. These methods should also use `__getitem__` or other methods instead of access to `storage` which is implementation specific.

Also they should be annotated like the rest of the code here.

Reply:
OK